decentraland/kernel
# feat: implement @dcl/rpc

# What? 
https://github.com/decentraland/sdk/issues/313
Replace the `decentraland-rpc` usage with `@dcl/rpc` in scenes.

# Why? 
`decentraland-rpc` doesn't support binary messages, an important feature to optimize the protocol and to implement the new ECS7.

# How to test
It's possible to test in every environment, just add `&kernel-branch=feat/implement-@dcl-rpc` in the URL. The behavior should be the same as if the query parameter isn't added.

The overhead added (temporary) is:
- JSON.stringify and JSON.parse in the events payload, the most case used is the position report and avatar scenes.
- Each engine scene message was JSON-serialized, now it's JSON-serialized and then protocolbuffer-serialized. In other words, one layer of serialization is **temporary** added, so this can introduce some delays or make it heavier.

Some environment examples:
- Prod https://play.decentraland.org/?position=0,0&kernel-branch=feat/implement-@dcl-rpc
- Zone with ROPSTEN https://play.decentraland.zone/?position=0,0&kernel-branch=feat/implement-@dcl-rpc&NETWORK=ropsten
- SDK test scenes https://sdk-test-scenes.decentraland.io/?kernel-branch=feat/implement-@dcl-rpc
- Your local preview https://127.0.0.1:8000/?kernel-branch=feat/implement-@dcl-rpc

**You have two paths in each environment, as a guest and with a wallet.**

# How the code was changed

## @dcl/rpc host implementation
**Path: `packages/world`**

> In the previous implementation, kernel fed the APIs with `system.getAPIInstance(XXXX)` getting the class implemented and modifying some variable members. 

Now, the `SceneWorker` creates the context with all the data before starting the system. This context is passed in all service methods and it can be patched by accessing the `SceneWorker` object and modifying the `rpcContext` member.

## Scene runtime implementation
**Path: `packages/scene-system`**
The `SceneRuntime.ts` was rewritten, now the scene runtime consumes `event` from the EngineAPI with a pushable channel, instead of native on-emit messages from the worker (decentraland-rpc really).

## APIs 
- Path: `packages/shared/apis`
The changed line counter is significant because of this.

1. All APIs were transcribed to `.proto` as RPC services. 
2. All API calls were modified because the request and response data are only one object instead of multiple parameters, undefined responses or void responses, etc. 
3. To maintain the backward compatibility, most of the API-clients preserved the response and request data structure except EngineAPI (now the events are pushed to the channel, and this channel streamed)
  a. The event payloads now have JSON.stringify on the host-side and JSON.parse on the worker-side, this can introduce some delays  
5. SignedFetch is another modified API, the request object was trimmed to only `method`, `body`, and `headers`.


# Known issues (X indicates that it's solved)
- [x] Delayed event receiving: sometimes it seems some events are enqueued and the delivery delays
- [x] When close the worker, the host throw the error that is expecting the ACK message:
![Screenshot_20220606_191024](https://user-images.githubusercontent.com/8042536/172257603-806f785f-5ddc-4fed-a1b1-7ec012666221.png)
- [x] Fix avatar tests
- [x] Fix circular dependency (With StatefulWorker)
